### DISCUSSION TITLE
`BackendApplication#onStop` is not called on electron app quit. Is it a debug glitch, bug, or a missing feature?

### DISCUSSION THREAD

**Discussion by kittaakos** [eclipse-theia/theia#12135] [Q&A]
Hi there; when I quit a Theia-based electron app, or the Theia electron example started from sources by closing the last window with the `X`, the [`BackendApplication#onStop`](https://github.com/eclipse-theia/theia/blob/b28c8b9d543e7df6c5d2bc524d40c50414df7e4d/packages/core/src/node/backend-application.ts#L296) is not always called. When the electron app runs in cluster mode, and the Theia backend is forked from the electron main, `onStop` is not called from the electron app. When [`--no-cluster`](https://github.com/eclipse-theia/theia/blob/b28c8b9d543e7df6c5d2bc524d40c50414df7e4d/.vscode/launch.json#L33) arg is set, it works.

```
2023-02-01T14:13:35.812Z root INFO >>> Stopping backend contributions...
2023-02-01T14:13:35.812Z root INFO <<< All backend contributions have been stopped.
```

When the backend app process stops, it stops the contributions. This is the happy path, and it works fine if `--no-cluster` is set:

https://github.com/eclipse-theia/theia/blob/b28c8b9d543e7df6c5d2bc524d40c50414df7e4d/packages/core/src/node/backend-application.ts#L177

But in a bundled app, the backend process is forked from the electron main. The electron app kills the backend process on quit with the [default `'SIGTERM'` signal](https://nodejs.org/api/child_process.html#subprocesskillsignal):

https://github.com/eclipse-theia/theia/blob/b28c8b9d543e7df6c5d2bc524d40c50414df7e4d/packages/core/src/electron-main/electron-main-application.ts#L499

Currently, the backend app does not do any cleanup on `'SIGTERM'`:

https://github.com/eclipse-theia/theia/blob/b28c8b9d543e7df6c5d2bc524d40c50414df7e4d/packages/core/src/node/backend-application.ts#L181

https://github.com/eclipse-theia/theia/blob/b28c8b9d543e7df6c5d2bc524d40c50414df7e4d/packages/core/src/node/backend-application.ts#L173-L175

For example, I want to delete a file on quit synchronously, but it does not work because the backend contributions are not stopped. 

https://github.com/eclipse-theia/theia/blob/b28c8b9d543e7df6c5d2bc524d40c50414df7e4d/packages/core/src/node/backend-application.ts#L91-L97

Has somebody experienced the same? Is it a bug, or am I doing something wrong? Thanks!

**Comment by msujew** [comment]
I haven't confirmed it, but from your description, it looks like a bug to me.

**Reply by dmm9** [reply]
@msujew Could you confirm this is a bug? See also #14216

**Comment by dmm9** [comment]
I am experiencing the same, `onStop` is not being called. I have tried both `BackendApplicationContribution` and `ElectronMainApplicationContribution`. Using theia 1.49.1.

@kittaakos did you find a solution?

**Reply by dankeboy36** [reply]
> did you find a solution?

@dmm9, the [Arduino IDE](https://github.com/arduino/arduino-ide) has patched this default Theia behavior in this commit: https://github.com/arduino/arduino-ide/commit/d68bc4abdb5d5a9da3b547bee7bba15dfadbc802#diff-54951a7f6fc1d8e5bca375af44a6874aa403c6ab0942031b7126743fc7d5ab66.

**Comment by dmm9** [comment]
Bug #14216
